homeassistant/components/sensor/__init__.py

```diff
@@ -545,6 +553,8 @@ def from_dict(cls, restored: dict[str, Any]) -> SensorExtraStoredData | None:
                 native_value = dt_util.parse_datetime(native_value["isoformat"])
             elif type_ == "<class 'datetime.date'>":
                 native_value = dt_util.parse_date(native_value["isoformat"])
+            elif type_ == "<class 'decimal.Decimal'>":
+                native_value = Decimal(native_value["decimal_str"])
```

Comment:
This will throw `decimal.InvalidOperation` if the stored data is invalid, let's catch that and return `None` if caught.

Reply:
I initially had that exception handling, but I realised that having a corrupt storage file would have been a bigger issue then just the decimal values.

I will nonetheless workout code to address this

Reply:
exception handling and tests added